I don't understand how this exception is related to the change in the PR:

```
    async def cancel(self) -> None:
>       async with cleanup_tasks(self.asyncio_task):
E       AttributeError: 'FunctionTask' object has no attribute 'asyncio_task'

self       = <async_service.asyncio.FunctionTask object at 0x7fadac32ea90>

async_service/asyncio.py:46: AttributeError
========================================================================================================= slowest 10 test durations ==========================================================================================================
0.46s call     tests-asyncio/test_asyncio_based_service.py::test_asyncio_service_with_too_many_children

```

--- a/async_service/__init__.py
+++ b/async_service/__init__.py
@@ -5,7 +5,11 @@ from .asyncio import (  # noqa: F401
     external_api as external_asyncio_api,
 )
 from .base import Service, as_service  # noqa: F401
-from .exceptions import DaemonTaskExit, LifecycleError  # noqa: F401
+from .exceptions import (  # noqa: F401
+    DaemonTaskExit,
+    LifecycleError,
+    TooManyChildrenException,
+)
 from .trio import (  # noqa: F401
     TrioManager,
     background_trio_service,

--- a/async_service/base.py
+++ b/async_service/base.py
@@ -1,5 +1,6 @@
 from abc import abstractmethod
 import asyncio
+from collections import Counter
 import logging
 import sys
 from typing import (
@@ -25,9 +26,11 @@ from .abc import (
     TaskAPI,
     TaskWithChildrenAPI,
 )
-from .exceptions import DaemonTaskExit, LifecycleError
+from .exceptions import DaemonTaskExit, LifecycleError, TooManyChildrenException
 from .stats import Stats, TaskStats
 from .typing import EXC_INFO, AsyncFn
+
+MAX_CHILDREN_TASKS = 1000
 
 
 class Service(ServiceAPI):
@@ -283,20 +286,34 @@ class BaseManager(InternalManagerAPI):
             )
             return
 
-        if task.parent is None:
+        self._add_child_task(task.parent, task)
+        self._total_task_count += 1
+
+        self._schedule_task(task)
+
+    def _add_child_task(
+        self, parent: Optional[TaskWithChildrenAPI], task: TaskAPI
+    ) -> None:
+        if parent is None:
+            all_children = self._root_tasks
+        else:
+            all_children = parent.children
+
+        if len(all_children) > MAX_CHILDREN_TASKS:
+            task_counter = Counter(map(str, all_children))
+            raise TooManyChildrenException(
+                f"Tried to add more than {MAX_CHILDREN_TASKS} child tasks."
+                f" Most common tasks: {task_counter.most_common(10)}"
+            )
+
+        if parent is None:
             if self._verbose:
                 self.logger.debug("%s: running root task %s", self, task)
             self._root_tasks.add(task)
         else:
             if self._verbose:
-                self.logger.debug(
-                    "%s: %s running child task %s", self, task.parent, task
-                )
-            task.parent.add_child(task)
-
-        self._total_task_count += 1
-
-        self._schedule_task(task)
+                self.logger.debug("%s: %s running child task %s", self, parent, task)
+            parent.add_child(task)
 
     async def _run_and_manage_task(self, task: TaskAPI) -> None:
         if self._verbose:
@@ -313,10 +330,7 @@ class BaseManager(InternalManagerAPI):
                         new_parent = task.parent
                         for child in task.children:
                             child.parent = new_parent
-                            if new_parent is None:
-                                self._root_tasks.add(child)
-                            else:
-                                new_parent.add_child(child)
+                            self._add_child_task(new_parent, child)
                             self.logger.debug(
                                 "Daemon %s left child task (%s) behind, reassigning it to %s",
                                 task,

--- a/async_service/exceptions.py
+++ b/async_service/exceptions.py
@@ -20,3 +20,12 @@ class DaemonTaskExit(ServiceException):
     """
 
     pass
+
+
+class TooManyChildrenException(ServiceException):
+    """
+    Raised when a service adds too many children. It is a sign of task leakage
+    that needs to be prevented.
+    """
+
+    pass

--- a/tests-asyncio/test_asyncio_based_service.py
+++ b/tests-asyncio/test_asyncio_based_service.py
@@ -8,6 +8,7 @@ from async_service import (
     DaemonTaskExit,
     LifecycleError,
     Service,
+    TooManyChildrenException,
     as_service,
     background_asyncio_service,
 )
@@ -561,3 +562,32 @@ async def test_asyncio_service_with_try_finally_cleanup_with_await():
         assert not service.cleanup_up
         manager.cancel()
     assert service.cleanup_up
+
+
+@pytest.mark.asyncio
+async def test_asyncio_service_with_too_many_children():
+    class ServiceTest(Service):
+        async def run(self):
+            await self.manager.wait_finished()
+
+        async def add_child(self):
+            event = asyncio.Event()
+            self.manager.run_task(self._notify_and_sleep, event)
+            await event.wait()
+
+        async def _notify_and_sleep(self, event):
+            event.set()
+            await asyncio.sleep(100)
+
+    service = ServiceTest()
+
+    # TODO same test but with service/task DAG
+    async with background_asyncio_service(service) as manager:
+        for child in range(1000):
+            await service.add_child()
+
+        assert manager.is_running
+        assert not manager.did_error
+
+        with pytest.raises(TooManyChildrenException):
+            await service.add_child()

